test/evp_test.c

```diff
@@ -74,6 +74,27 @@ static int find_key(EVP_PKEY **ppk, const char *name, KEY_LIST *lst);
 
 static int parse_bin(const char *value, unsigned char **buf, size_t *buflen);
 
+/*
+ * Compare two memory regions for equality, returning zero if they differ.
+ * However, if there is expected to be an error and the actual error
+ * matches then the memory is expected to be different so handle this
+ * case without producing unnecessary test framework output.
+ */
+static int memory_err_compare(EVP_TEST *t, const char *err,
+                              const void *expected, size_t expected_len,
+                              const void *got, size_t got_len)
+{
+    int r;
+
+    if (t->expected_err != NULL && strcmp(t->expected_err, err) == 0)
+        r = !TEST_mem_ne(expected, expected_len, got, got_len);
+    else
+        r = TEST_mem_eq(expected, expected_len, got, got_len);
```

Comment:
All error messages will be logged with line number 90 or 92.

Reply:
I wasn't too concerned about this.  The output that `evp_test` already produces uniquely determines the failure line -- search for the error or look at the kind of algorithm being run.  Either describes it.

However, I'm dubious about tests using the internals of testutil.  The macros are the defined interface, the underlying functions aren't intended to be called directly.

Does it make sense to add TEST_mem_eq_err and friends to the framework?  I.e. tests which are expecting an error and produce output when there isn't one?  I'm not convinced they'd be sufficiently useful.

Reply:
I checked it again, you are right. The improvement of the output gained by my patch is very small. So forget about my suggestion.